src/preprocessing.ml

```diff
@@ -45,52 +55,39 @@ end = struct
         | None  , Some _ -> scope_for_key
         | None  , None   -> None)
     in
-    let scope_key =
-      match scope_for_key with
-      | None -> ""
-      | Some name ->
-        match Dune_project.Name.to_encoded_string name with
-        | "" -> assert false
-        | s -> s
-    in
-    let key =
-      (scope_key
-       :: List.map libs ~f:(fun lib -> Lib.name lib |> Lib_name.to_string))
-      |> String.concat ~sep:"\000"
-      |> Digest.string
-    in
-    let full_key = (ctx.name, key) in
-    begin
-      match Hashtbl.find reverse_table full_key with
-      | None ->
-        Hashtbl.add reverse_table full_key libs
-      | Some libs' ->
-        match List.compare libs libs' ~compare:(fun a b ->
-          Int.compare (Lib.unique_id a) (Lib.unique_id b)) with
-        | Eq -> ()
-        | _ ->
-          let string_of_libs libs =
-            String.concat ~sep:", " (List.map libs ~f:(fun lib ->
-              sprintf "%s (in %s)"
-                (Lib_name.to_string (Lib.name lib))
-                (Path.to_string_maybe_quoted (Lib.src_dir lib))))
-          in
-          die "Hash collision between set of ppx drivers:\n\
-               - cache : %s\n\
-               - fetch : %s\n\
-               context: %s"
-            (string_of_libs libs)
-            (string_of_libs libs')
-            ctx.name
-    end;
-    key
-
-  let decode (ctx : Context.t) key =
-    match Hashtbl.find reverse_table (ctx.name, key) with
-    | Some libs -> libs
+    { pps; scope }
+
+  let encode x =
+    let y = Digest.string (Marshal.to_string x []) in
+    match Hashtbl.find reverse_table y with
+    | None ->
+      Hashtbl.add reverse_table y x;
+      y
+    | Some x' ->
+      if x = x' then
+        y
+      else begin
+        let to_string { pps; scope } =
+          let s = String.enumerate_and (List.map pps ~f:Lib_name.to_string) in
+          match scope with
+          | None -> s
+          | Some scope ->
+            sprintf "%s (in project: %s)" s
+              (Dune_project.Name.to_string_hum scope)
+        in
+        die "Hash collision between set of ppx drivers:\n\
```

Comment:
Do you think it's feasible to add a test case covering this failure mode?

Reply:
The function that is being reverse is simpler now, before it was `Lib.t list -> Digest.t`, while now it's just `string list * string option -> Digest.t`, so the same results can be reused between scopes/build contexts/builds